### PR TITLE
Add cuda-samples

### PR DESCRIPTION
**Pull request by xnox**
trying to add cuda-samples, and questioning if this is interesting or not......

.... given that it seems to be arch restricted to amd64|arm64 and i'm failing to figure out how to make arch-specific only parts/apps.

Note that on unsupported arches the two new apps are just bin/false, let me know if that's ok, or if you want a wrapper that explains things.

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -24,9 +24,15 @@ description: |
...
     Test the Vulkan setup.
+  - bandwidthTest (MIT):
+    A sample NVIDIA CUDA application can be used to confirm CUDA operational status.
+    (needs nvidia-core22)
+  - deviceQuery (MIT):
+    Query NVIDIA CUDA hardware configuration. (needs nvidia-core22)
 
   This snap uses the graphics-core22 interface to gain support graphics hardware.
   It can be used to verify custom graphics-core22 provider implementations.
+  The CUDA specific apps require NVIDIA capable provider, such as nvidia-core22.
```
Comment (Saviq) [inline snap/snapcraft.yaml:35]:
```suggestion
  The CUDA specific apps require Nvidia capable provider, such as nvidia-core22.
```

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -133,6 +153,38 @@ parts:
     stage-packages:
       - inotify-tools
 
+  cuda-samples:
+    source: https://github.com/NVIDIA/cuda-samples
+    source-tag: v11.5
+    source-type: git
```
Comment (Saviq) [inline snap/snapcraft.yaml:159]: Should we try and just build latest? It's 12.1 by now - and I imagine we'll be falling behind quickly if we stick to a version. I suppose could subscribe to release notifications there…

```suggestion
```

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -133,6 +153,38 @@ parts:
...
+    source-depth: 1
+    plugin: nil
+    build-packages:
+      - build-essential
+      - on amd64:
+        - libnvidia-compute-515-server
+        - nvidia-cuda-toolkit
+      - on arm64:
+        - libnvidia-compute-515-server
+        - nvidia-cuda-toolkit
```
Comment (Saviq) [inline snap/snapcraft.yaml:169]:
```suggestion
    build-packages:
      - on amd64,arm64:
        - build-essential
        - libnvidia-compute-515-server
        - nvidia-cuda-toolkit
```

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -133,6 +153,38 @@ parts:
...
+    override-build: |
```
Comment (Saviq) [inline snap/snapcraft.yaml:170]:
```suggestion
    override-build: |
      mkdir -p $SNAPCRAFT_PART_INSTALL/usr/bin/
```

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -133,6 +153,38 @@ parts:
...
+      case $CRAFT_TARGET_ARCH in
+      amd64|arm64)
+      make -C Samples/bandwidthTest CUDA_PATH=/usr/lib/nvidia-cuda-toolkit
+      make -C Samples/deviceQuery CUDA_PATH=/usr/lib/nvidia-cuda-toolkit
+      mkdir -p $SNAPCRAFT_PART_INSTALL/usr/bin/
```
Comment (Saviq) [inline snap/snapcraft.yaml:175]:
```suggestion
```

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -133,6 +153,38 @@ parts:
...
+      cp bin/*/linux/release/* $SNAPCRAFT_PART_INSTALL/usr/bin/
+      ;;
+      *)
+      mkdir -p $SNAPCRAFT_PART_INSTALL/usr/bin/
```
Comment (Saviq) [inline snap/snapcraft.yaml:179]:
```suggestion
```

File: snap/snapcraft.yaml [hunk 7]
```diff
@@ -133,6 +153,38 @@ parts:
...
+      cp /bin/false $SNAPCRAFT_PART_INSTALL/usr/bin/bandwidthTest
+      cp /bin/false $SNAPCRAFT_PART_INSTALL/usr/bin/deviceQuery
```
Comment (Saviq) [inline snap/snapcraft.yaml:181]: Could these be symlinks instead?

File: snap/snapcraft.yaml (hunk 6 above)
Comment (Saviq) [inline snap/snapcraft.yaml:179]:
```suggestion
```

File: snap/snapcraft.yaml [hunk 8]
```diff
@@ -24,9 +24,15 @@ description: |
     Dump information about the available Vulkan setup.
   - vkcube (Apache-2.0):
     Test the Vulkan setup.
+  - bandwidthTest (MIT):
+    A sample NVIDIA CUDA application can be used to confirm CUDA operational status.
```
Comment (Saviq) [inline snap/snapcraft.yaml:28]:
```suggestion
    A sample Nvidia CUDA application can be used to confirm CUDA operational status.
```

File: snap/snapcraft.yaml [hunk 9]
```diff
@@ -24,9 +24,15 @@ description: |
...
+    (needs nvidia-core22)
+  - deviceQuery (MIT):
+    Query NVIDIA CUDA hardware configuration. (needs nvidia-core22)
```
Comment (Saviq) [inline snap/snapcraft.yaml:31]:
```suggestion
    Query Nvidia CUDA hardware configuration. (needs nvidia-core22)
```

File: snap/snapcraft.yaml [hunk 10]
```diff
@@ -24,9 +24,15 @@ description: |
     Dump information about the available Vulkan setup.
   - vkcube (Apache-2.0):
     Test the Vulkan setup.
+  - bandwidthTest (MIT):
+    A sample NVIDIA CUDA application can be used to confirm CUDA operational status.
+    (needs nvidia-core22)
+  - deviceQuery (MIT):
```
Comment (Saviq) [inline snap/snapcraft.yaml:30]: Should we call these `nvidia-` so it's clear?

**Review by Saviq — CHANGES_REQUESTED** [review]
Looks fine overall, nitpicks inline.

File: snap/snapcraft.yaml (hunk 1 above)
Comment (xnox) [reply snap/snapcraft.yaml:35]: NVIDIA should always be refered to in all-caps, as per their https://www.nvidia.com/en-us/about-nvidia/legal-info/logo-brand-usage/ guidelines (despite small-caps wordmark in the logo assets).

File: snap/snapcraft.yaml (hunk 2 above)
Comment (xnox) [reply snap/snapcraft.yaml:159]: This was the highest version at the time, compatible with the -server version of drivers we used at the time, compatible with focal & jammy. I can check if 12.x series are usable for us now, but they weren't compatible before when we were on older generation of driver =/

File: snap/snapcraft.yaml (hunk 3 above)
Comment (xnox) [reply snap/snapcraft.yaml:169]: is that valid?

From https://snapcraft.io/docs/snapcraft-advanced-grammar

> The body of the on clause is taken into account if every (AND, not OR) selector is true for the build environment. Currently the only selectors supported are host architectures (e.g. amd64).

something cannot be simulntationlsy on amd64 AND on arm64, meaning none of those get installed?

File: snap/snapcraft.yaml (hunk 7 above)
Comment (xnox) [reply snap/snapcraft.yaml:181]: review tools usually rejects broken symlinks or symlinks pointing outside of the snap. Unless this has changed.....

File: snap/snapcraft.yaml (hunk 10 above)
Comment (xnox) [reply snap/snapcraft.yaml:30]: no preference, i guess we could. Both of those names are widely known / associated with CUDA, in this cammelCase spelling.

File: snap/snapcraft.yaml (hunk 1 above)
Comment (Saviq) [reply snap/snapcraft.yaml:35]: ACK

File: snap/snapcraft.yaml (hunk 10 above)
Comment (Saviq) [reply snap/snapcraft.yaml:30]: Right, but if someone comes to g-t-t without CUDA knowledge, they may try them. I'd prefer `nvidia-*`, or maybe even `cuda-*`?

File: snap/snapcraft.yaml (hunk 2 above)
Comment (Saviq) [reply snap/snapcraft.yaml:159]: Your call.

File: snap/snapcraft.yaml (hunk 3 above)
Comment (Saviq) [reply snap/snapcraft.yaml:169]: I should RTFM. Thanks.

File: snap/snapcraft.yaml (hunk 7 above)
Comment (Saviq) [reply snap/snapcraft.yaml:181]: Even better, make these `echo 'This only works on amd64 and arm64'; false`?

**Pull request closed without merging**